boomstream.py

```diff
@@ -180,91 +197,123 @@ def get_aes_key(self, xmedia_ready):
+            run_bash(f'curl -s "{line}" | openssl aes-128-cbc -K "{hex_key}" -iv "{iv}" -d > {outf}')
             i += 1
+        return filenames
 
-    def merge_chunks(self, key):
+    def merge_chunks(self, filenames, key, expected_result_duration):
         """
         Merges all chunks into one file and encodes it to MP4
         """
         print("Merging chunks...")
-        os.system("cat %s/*.ts > %s.ts" % (key, key,))
+        run_bash(f"cat {' '.join(filenames)} > {output_path(key)}.ts")
         print("Encoding to MP4")
-        os.system('ffmpeg -i %s.ts -c copy "%s".mp4' % (key, self.get_title(),))
+        run_bash(f'ffmpeg -nostdin -y -i {output_path(key)}.ts -c copy {output_path(key)}.mp4')
+
+        result_format = run_bash(f'ffprobe -i {output_path(key)}.mp4 -show_format')
+        result_duration = float([line[len("duration="):] for line in result_format.split('\n') if line.startswith("duration=")][0])
+        print(f"Result duration: {result_duration:.2f}")
+        print(f"Expected duration: {expected_result_duration:.2f}")
+        if abs(result_duration - expected_result_duration) > 2:
+            raise ValueError(f"unexpected result duration: {expected_result_duration:.2f} != {result_duration:.2f}")
+
+        ensure_folder_exists(output_path("results"))
+        result_filename = output_path(os.path.join("results", f"{valid_filename(self.get_title())}.mp4"))
+        os.rename(f'{output_path(key)}.mp4', result_filename)
 
     def get_title(self):
         return self.config['entity']['title']
 
+    def get_access_cookies(self):
+        pin = self.args.pin
+        if pin is None:
+            return {}
+        r = requests.post("https://play.boomstream.com/api/subscriptions/recovery",
+                          headers={'content-type': 'application/json;charset=UTF-8'},
+                          data=f'{{"entity":"{self.args.entity}","code":"{pin}"}}')
+        response = json.loads(r.text)
+        if "data" not in response or "cookie" not in response["data"]:
+            if "errors" not in response or "code" not in response:
+                raise ValueError(f"unexpected response on authorization: {r.text}")
+            else:
+                raise ValueError(f"authorization failed: {response['code']} {response['errors']}")
+        cookie = response["data"]["cookie"]
+        return {cookie["name"]: cookie["value"]}
+
     def run(self):
-        if self.args.use_cache and os.path.exists('result.html'):
-            page = open('result.html').read()
+        ensure_folder_exists(OUTPUT_PATH)
+
+        cookies = self.get_access_cookies()
+
+        result_path = output_path('result.html')
+
+        if self.args.use_cache and os.path.exists(result_path):
+            page = open(result_path).read()
         else:
-            r = requests.get(self.args.url, headers=headers)
+            r = requests.get(f'https://play.boomstream.com/{self.args.entity}', headers=headers, cookies=cookies)
```

Comment:
I think for user would be easier to copy the URL from browser command line than extracting and copying `entity` argument manually.